[topi][relay] add operation tan to TVM

notoraptor opened the pull request:
Hi! This pull request adds operation tan to TVM. The main goal is to make it available in relay module.

kevinthesun commented on python/tvm/relay/op/_tensor.py:
```diff
@@ -27,6 +27,7 @@
 
 
 register_broadcast_schedule("log")
+register_broadcast_schedule("tan")
```
Add shape function for tan as well?

notoraptor replied:
Hi ! What do you mean by adding shape function for tan ?

kevinthesun replied:
https://github.com/apache/incubator-tvm/blob/b6893017d964f531e7f1600faaec4e3817fe89e2/python/tvm/relay/op/_tensor.py#L185-L217

kevinthesun commented:
Minor comments. Also can you rebase master?

notoraptor commented:
@kevinthesun Hi ! Thanks!

In order to fix failed test (precision error), I am trying to update the `tan` implementation in llvm target (by replacing `sin/cos` with a direct call to llvm's `tan` functions). But it currently generate other issues, so I am working on it.

I can rebase now, but I guess test will fail again.

notoraptor commented:
Rebased and updated to try to fix failing test.

notoraptor commented:
Hi @kevinthesun I rebased and added shape function for tan in latest commit.

I still have a mismatch error in `tests/python/frontend/tensorflow/test_forward.py::test_forward_tan`. It's hard to debug because I cannot reproduce it on my computer. But I am still working on it.

notoraptor commented:
@kevinthesun Tests passed ! I implemented tan in LLVM using `sin/cos` and I fixed CUDA implementation by using `tanf` instead of `__tanf` for float32.

kevinthesun commented:
@notoraptor Thanks!

tqchen commented:
NOTE: the git commit message messed up today due to github issue https://discuss.tvm.ai/t/github-issue-the-commit-author-is-wrong-since-today/5880/15 we should refrain from merging others' PR until tomorrow

kevinthesun commented:
@tqchen Do we need to do something for this PR?

tqchen commented:
See https://discuss.tvm.ai/t/response-to-github-author-attribution-bug/5909/2

tqchen commented:
Hi @notoraptor @kevinthesun The community has reach a concensus to correct the github author attribution problem by reverting the affected PR, and ask the contributor to resend the pull request.

See background here https://discuss.tvm.ai/t/response-to-github-author-attribution-bug/5909
It will need an coordindated effort
- first @kevinthesun first revert the PR
- then @notoraptor to resent the PR to the master

Please reply to the post to see if you are interested in doing so. As a community we value every contribution and would like to make things right. Once both of you confirms, we can go ahead and expedite the process

notoraptor commented:
Hi @tqchen ! OK for that.

Anyway, if it causes some problems, then we can just let things as it is, as my main goal was just to make `tan` available in relay !

Thanks !

tqchen commented:
cc @kevinthesun please send the reverting PR, then @notoraptor can send the patch
